From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Mon, 1 Aug 2022 22:50:29 -0400
Subject: [PATCH] Brigadier based command API


diff --git a/build.gradle.kts b/build.gradle.kts
index e950bce5047552bdd3f5664eb24ce290b0a06225..e043ccc13d9570ae18685ebfbddbb0481e416221 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -46,6 +46,7 @@ dependencies {
     implementation("org.ow2.asm:asm-commons:9.2")
     // Paper end
 
+    api("com.mojang:brigadier:1.0.18") // Paper, expose!
     compileOnly("org.apache.maven:maven-resolver-provider:3.8.5")
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
     compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..34e062202b52a450951fc0eeac9a0813a3b42120
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java
@@ -0,0 +1,86 @@
+package io.papermc.paper.command.brigadier;
+
+import com.mojang.brigadier.CommandDispatcher;
+import com.mojang.brigadier.builder.ArgumentBuilder;
+import com.mojang.brigadier.tree.RootCommandNode;
+import org.bukkit.Bukkit;
+import org.bukkit.plugin.Plugin;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class CommandBuilder extends ArgumentBuilder<CommandSourceStack, CommandBuilder> {
+
+    private final Plugin plugin;
+    private final String literal;
+
+    private String description;
+    private List<String> aliases = List.of();
+
+    protected CommandBuilder(Plugin plugin, String literal) {
+        this.literal = literal;
+        this.plugin = plugin;
+    }
+
+    public static CommandBuilder of(Plugin plugin, String name) {
+        return new CommandBuilder(plugin, name);
+    }
+
+    public CommandBuilder aliases(String... aliases) {
+        this.aliases = Arrays.asList(aliases);
+        return this.getThis();
+    }
+
+    public String getDescription() {
+        return this.description;
+    }
+
+    public CommandBuilder description(String description) {
+        this.description = description;
+        return this;
+    }
+
+    public Plugin getPlugin() {
+        return this.plugin;
+    }
+
+    @Override
+    protected CommandBuilder getThis() {
+        return this;
+    }
+
+    @Override
+    public PluginCommandNode build() {
+        // See: LiteralArgumentBuilder
+        return new PluginCommandNode(this.literal, this);
+    }
+
+    public void register() {
+        this.register(Bukkit.getServer().getCommandDispatcher());
+    }
+
+    public void register(CommandDispatcher<CommandSourceStack> dispatcher) {
+        String identifier = this.plugin.getDescription().getName();
+        RootCommandNode<CommandSourceStack> dispatcherRoot = dispatcher.getRoot();
+        PluginCommandNode root = this.build();
+
+        dispatcherRoot.addChild(root);
+        // Add root redirect
+        this.registerRedirect(dispatcherRoot, root, identifier + ":" + this.literal);
+
+        // Add aliases
+        for (String alias : this.aliases) {
+            this.registerRedirect(dispatcherRoot, root, alias);
+            this.registerRedirect(dispatcherRoot, root, identifier + ":" + alias);
+        }
+    }
+
+    private void registerRedirect(RootCommandNode<CommandSourceStack> dispatcherRoot, PluginCommandNode node, String name) {
+        PluginCommandNode aliasRoot = CommandBuilder.of(this.plugin, name)
+            .description(this.description)
+            .redirect(node)
+            .build();
+
+        dispatcherRoot.addChild(aliasRoot);
+    }
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java b/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java
new file mode 100644
index 0000000000000000000000000000000000000000..3d2ca2d0e092c1aa15adf7a32da8567c614aa43f
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java
@@ -0,0 +1,23 @@
+package io.papermc.paper.command.brigadier;
+
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.command.CommandSender;
+import org.bukkit.entity.Entity;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+@ApiStatus.NonExtendable
+public interface CommandSourceStack {
+
+    @Nullable
+    Entity getBukkitEntity();
+
+    @Nullable
+    World getBukkitWorld();
+
+    @Nullable
+    Location getBukkitLocation();
+
+    CommandSender getBukkitSender();
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java
new file mode 100644
index 0000000000000000000000000000000000000000..839927c90a095674a332a8f61e3c279c4c08129a
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java
@@ -0,0 +1,18 @@
+package io.papermc.paper.command.brigadier;
+
+import com.mojang.brigadier.Message;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.serializer.ComponentSerializer;
+
+public interface MessageComponentSerializer extends ComponentSerializer<Component, Component, Message> {
+
+    /**
+     * A component serializer that converts it to a brigadier compatible type.
+     *
+     * @return serializer instance
+     */
+    static MessageComponentSerializer message() {
+        return MessageComponentSerializerHolder.PROVIDER;
+    }
+
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerHolder.java b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerHolder.java
new file mode 100644
index 0000000000000000000000000000000000000000..61ee6e72643dbd1e46dce362adddc9936631df95
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerHolder.java
@@ -0,0 +1,10 @@
+package io.papermc.paper.command.brigadier;
+
+import java.util.ServiceLoader;
+
+class MessageComponentSerializerHolder {
+
+    public static final MessageComponentSerializer PROVIDER = ServiceLoader.load(MessageComponentSerializer.class)
+        .findFirst()
+        .orElseThrow();
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/PluginCommandNode.java b/src/main/java/io/papermc/paper/command/brigadier/PluginCommandNode.java
new file mode 100644
index 0000000000000000000000000000000000000000..d1809cfe836129e84ab8a1fe780fd2fe864ebd9e
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/PluginCommandNode.java
@@ -0,0 +1,34 @@
+package io.papermc.paper.command.brigadier;
+
+import com.mojang.brigadier.tree.CommandNode;
+import com.mojang.brigadier.tree.LiteralCommandNode;
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Internal
+public class PluginCommandNode extends LiteralCommandNode<CommandSourceStack> {
+
+    private final Plugin plugin;
+    private final String description;
+
+    public PluginCommandNode(String literal, CommandBuilder builder) {
+        super(
+            literal, builder.getCommand(), builder.getRequirement(),
+            builder.getRedirect(), builder.getRedirectModifier(), builder.isFork()
+        );
+        this.plugin = builder.getPlugin();
+        this.description = builder.getDescription();
+
+        for (CommandNode<CommandSourceStack> argument : builder.getArguments()) {
+            this.addChild(argument);
+        }
+    }
+
+    public Plugin getPlugin() {
+        return this.plugin;
+    }
+
+    public String getDescription() {
+        return this.description;
+    }
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/Message.java b/src/main/java/io/papermc/paper/command/brigadier/argument/Message.java
new file mode 100644
index 0000000000000000000000000000000000000000..6c922dfb11c3513b39e632bec5f136c788e9c017
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/argument/Message.java
@@ -0,0 +1,7 @@
+package io.papermc.paper.command.brigadier.argument;
+
+public interface Message {
+
+    String getSignedContent();
+
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArgumentProvider.java b/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArgumentProvider.java
new file mode 100644
index 0000000000000000000000000000000000000000..a693760babf984765598cf7d3f24c3e9d2f1df49
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArgumentProvider.java
@@ -0,0 +1,22 @@
+package io.papermc.paper.command.brigadier.argument;
+
+import com.mojang.brigadier.arguments.ArgumentType;
+import org.bukkit.NamespacedKey;
+import org.bukkit.block.BlockState;
+import org.bukkit.block.data.BlockData;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.ApiStatus;
+
+import java.util.UUID;
+
+@ApiStatus.Internal
+interface VanillaArgumentProvider {
+
+    ArgumentType<ItemStack> itemStack();
+
+    ArgumentType<BlockState> blockState();
+
+    ArgumentType<Message> signedMessage();
+
+    ArgumentType<UUID> uuid();
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArguments.java b/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArguments.java
new file mode 100644
index 0000000000000000000000000000000000000000..e1cc17aa4feccf6ebfd31235827a200623020a2e
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArguments.java
@@ -0,0 +1,33 @@
+package io.papermc.paper.command.brigadier.argument;
+
+import com.mojang.brigadier.arguments.ArgumentType;
+import org.bukkit.block.BlockState;
+import org.bukkit.inventory.ItemStack;
+
+import java.util.ServiceLoader;
+import java.util.UUID;
+
+public final class VanillaArguments {
+
+    private static final VanillaArgumentProvider PROVIDER = ServiceLoader.load(VanillaArgumentProvider.class)
+        .findFirst()
+        .orElseThrow();
+
+    public static ArgumentType<ItemStack> itemStack() {
+        return PROVIDER.itemStack();
+    }
+
+    public static ArgumentType<BlockState> blockState() {
+        return PROVIDER.blockState();
+    }
+
+    public static ArgumentType<UUID> uuid() {
+        return PROVIDER.uuid();
+    }
+
+    public static ArgumentType<Message> signedMessage() {
+        return PROVIDER.signedMessage();
+    }
+
+
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/WrapperArgumentType.java b/src/main/java/io/papermc/paper/command/brigadier/argument/WrapperArgumentType.java
new file mode 100644
index 0000000000000000000000000000000000000000..9ed29557432cf155ba7cdac2ab6f50b31e1851be
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/argument/WrapperArgumentType.java
@@ -0,0 +1,72 @@
+package io.papermc.paper.command.brigadier.argument;
+
+import com.mojang.brigadier.StringReader;
+import com.mojang.brigadier.arguments.ArgumentType;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.brigadier.suggestion.Suggestions;
+import com.mojang.brigadier.suggestion.SuggestionsBuilder;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An argument type that wraps around another argument.
+ * The wrapper will be sent to the client.
+ *
+ * @param <T> type
+ * @param <B> base wrapping type
+ */
+public abstract class WrapperArgumentType<T, B> implements ArgumentType<T> {
+
+    private final ArgumentType<B> baseArgumentType;
+
+    public WrapperArgumentType(ArgumentType<B> baseArgumentType) {
+        this.baseArgumentType = baseArgumentType;
+    }
+
+    @Override
+    public T parse(StringReader reader) throws CommandSyntaxException {
+        return this.convert(this.baseArgumentType.parse(reader));
+    }
+
+    /**
+     * Converts the type from the wrapped argument to this argument.
+     * @param baseType wrapped argument provided value
+     * @return converted value
+     * @throws CommandSyntaxException if an exception occurs while parsing
+     */
+    public abstract T convert(B baseType) throws CommandSyntaxException;
+
+    /**
+     * Gets the wrapped argument.
+     * @return wrapped
+     */
+    public ArgumentType<B> getWrapped() {
+        return this.baseArgumentType;
+    }
+
+    /**
+     * Cannot be controlled by the server.
+     * Returned in cases where there are multiple arguments in the same node.
+     * This helps differentiate and tell the player what the possible inputs are.
+     *
+     * @return client set examples
+     */
+    @Override
+    public final Collection<String> getExamples() {
+        return this.baseArgumentType.getExamples();
+    }
+
+    /**
+     * Returns if the argument itself should handle sugggestion
+     * parsing instead of the wrapper.
+     *
+     * @return handles suggestions
+     */
+    public boolean handleSuggestions() {
+        return false;
+    }
+
+
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitBrigCommand.java b/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitBrigCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..5f45109ab27015b6b89dc5cc764b67edc726974f
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitBrigCommand.java
@@ -0,0 +1,57 @@
+package io.papermc.paper.command.brigadier.bukkit;
+
+import co.aikar.timings.Timing;
+import com.destroystokyo.paper.event.server.ServerExceptionEvent;
+import com.destroystokyo.paper.exception.ServerCommandException;
+import com.mojang.brigadier.Command;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import io.papermc.paper.command.brigadier.CommandSourceStack;
+import org.bukkit.Bukkit;
+import org.bukkit.Server;
+import org.bukkit.command.CommandException;
+import org.bukkit.command.CommandSender;
+
+import java.util.Arrays;
+
+class BukkitBrigCommand implements Command<CommandSourceStack> {
+
+    private final org.bukkit.command.Command command;
+    private final String literal;
+
+    BukkitBrigCommand(org.bukkit.command.Command command, String literal) {
+        this.command = command;
+        this.literal = literal;
+    }
+
+    @Override
+    public int run(CommandContext<CommandSourceStack> context) throws CommandSyntaxException {
+        Server server = Bukkit.getServer();
+
+        CommandSender sender = context.getSource().getBukkitSender();
+
+        // Plugins do weird things to workaround normal registration
+        if (command.timings == null) {
+            command.timings = co.aikar.timings.TimingsManager.getCommandTiming(null, command);
+        }
+
+        String[] args = org.apache.commons.lang3.StringUtils.split(context.getInput(), ' '); //  fix adjacent spaces (from console/plugins) causing empty array elements
+
+        try {
+            try (Timing ignored = command.timings.startTiming()) {
+                // Note: we don't return the result of target.execute as thats success / failure, we return handled (true) or not handled (false)
+                command.execute(sender, literal, Arrays.copyOfRange(args, 1, args.length));
+            }
+        } catch (CommandException ex) {
+            server.getPluginManager().callEvent(new ServerExceptionEvent(new ServerCommandException(ex, command, sender, args))); // Paper
+            throw ex;
+        } catch (Throwable ex) {
+            String msg = "Unhandled exception executing '" + context.getInput() + "' in " + command;
+            server.getPluginManager().callEvent(new ServerExceptionEvent(new ServerCommandException(ex, command, sender, args))); // Paper
+            throw new CommandException(msg, ex);
+        }
+
+        // return true as command was handled
+        return 1;
+    }
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitBrigSuggestionProvider.java b/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitBrigSuggestionProvider.java
new file mode 100644
index 0000000000000000000000000000000000000000..abbcf1047bb4539261fa824d9aae3917478395d8
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitBrigSuggestionProvider.java
@@ -0,0 +1,59 @@
+package io.papermc.paper.command.brigadier.bukkit;
+
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.brigadier.suggestion.SuggestionProvider;
+import com.mojang.brigadier.suggestion.Suggestions;
+import com.mojang.brigadier.suggestion.SuggestionsBuilder;
+import io.papermc.paper.command.brigadier.CommandSourceStack;
+import org.bukkit.Bukkit;
+import org.bukkit.ChatColor;
+import org.bukkit.Location;
+import org.bukkit.command.CommandException;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.logging.Level;
+
+class BukkitBrigSuggestionProvider implements SuggestionProvider<CommandSourceStack> {
+
+    public static final SuggestionProvider<CommandSourceStack> INSTANCE = new BukkitBrigSuggestionProvider();
+
+    @Override
+    public CompletableFuture<Suggestions> getSuggestions(CommandContext<CommandSourceStack> context, SuggestionsBuilder builder) throws CommandSyntaxException {
+        // Paper start
+        org.bukkit.command.CommandSender sender = context.getSource().getBukkitSender();
+        if (!(sender instanceof org.bukkit.entity.Player player)) {
+            return CompletableFuture.completedFuture(builder.build());
+        }
+
+        String message = builder.getInput();
+        List<String> results = null;
+        Location pos = context.getSource().getBukkitLocation();
+        try {
+            if (message.startsWith("/")) {
+                // Trim leading '/' if present (won't always be present in command blocks)
+                message = message.substring(1);
+            }
+            if (pos == null) {
+                results = Bukkit.getCommandMap().tabComplete(player, message);
+            } else {
+                results = Bukkit.getCommandMap().tabComplete(player, message, pos.clone());
+            }
+        } catch (CommandException ex) {
+            player.sendMessage(ChatColor.RED + "An internal error occurred while attempting to tab-complete this command");
+            Bukkit.getServer().getLogger().log(Level.SEVERE, "Exception when " + player.getName() + " attempted to tab complete " + message, ex);
+        }
+
+        // Paper end
+
+        // Defaults to sub nodes, but we have just one giant args node, so offset accordingly
+        builder = builder.createOffset(builder.getInput().lastIndexOf(' ') + 1);
+
+        for (String s : results) {
+            builder.suggest(s);
+        }
+
+        return builder.buildFuture();
+    }
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitCommandNode.java b/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitCommandNode.java
new file mode 100644
index 0000000000000000000000000000000000000000..8abaa489e537238ca848c976efa9e9df28c245c3
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitCommandNode.java
@@ -0,0 +1,37 @@
+package io.papermc.paper.command.brigadier.bukkit;
+
+import com.mojang.brigadier.arguments.StringArgumentType;
+import com.mojang.brigadier.builder.LiteralArgumentBuilder;
+import com.mojang.brigadier.builder.RequiredArgumentBuilder;
+import com.mojang.brigadier.tree.LiteralCommandNode;
+import io.papermc.paper.command.brigadier.CommandSourceStack;
+import org.bukkit.command.Command;
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Internal
+public class BukkitCommandNode extends LiteralCommandNode<CommandSourceStack> {
+
+    private final Command command;
+
+    public static BukkitCommandNode of(String name, Command command) {
+        BukkitBrigCommand bukkitBrigCommand = new BukkitBrigCommand(command, name);
+        BukkitCommandNode commandNode = new BukkitCommandNode(name, command, bukkitBrigCommand);
+        commandNode.addChild(
+            RequiredArgumentBuilder.<CommandSourceStack, String>argument("args", StringArgumentType.greedyString()).suggests(BukkitBrigSuggestionProvider.INSTANCE).executes(bukkitBrigCommand).build()
+        );
+
+        return commandNode;
+    }
+
+    private BukkitCommandNode(String literal, Command command, BukkitBrigCommand bukkitBrigCommand) {
+        super(
+            literal, bukkitBrigCommand, s -> command.testPermissionSilent(s.getBukkitSender()),
+            null, null, false
+        );
+        this.command = command;
+    }
+
+    public Command getBukkitCommand() {
+        return this.command;
+    }
+}
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 56e261efa654e4a6872ccea28f0461df13845d13..0c43387738ca3aceb414cb660964e090bec5979d 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2104,5 +2104,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * @return the potion brewer
      */
     @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer();
+
+    com.mojang.brigadier.CommandDispatcher<io.papermc.paper.command.brigadier.CommandSourceStack> getCommandDispatcher();
     // Paper end
 }
diff --git a/src/main/java/org/bukkit/command/CommandMap.java b/src/main/java/org/bukkit/command/CommandMap.java
index 864c263bbd4dd6dd7c37a74b39b1a40a884d0731..d5c10484497a091c46923e186f0a075cbbf33f05 100644
--- a/src/main/java/org/bukkit/command/CommandMap.java
+++ b/src/main/java/org/bukkit/command/CommandMap.java
@@ -5,6 +5,7 @@ import org.bukkit.Location;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+@Deprecated // Paper
 public interface CommandMap {
 
     /**
diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index b8623575b1c1b565560c2dd6438190716845a652..df9018c1783d6e44c471553b47ad6c256db40b0f 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -9,6 +9,8 @@ import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+
+import io.papermc.paper.command.brigadier.bukkit.BukkitCommandNode;
 import org.bukkit.Location;
 import org.bukkit.Server;
 import org.bukkit.command.defaults.BukkitCommand;
@@ -21,8 +23,30 @@ import org.bukkit.util.StringUtil;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+@Deprecated // Paper
 public class SimpleCommandMap implements CommandMap {
-    protected final Map<String, Command> knownCommands = new HashMap<String, Command>();
+    // Paper start
+    protected final Map<String, Command> knownCommands = new HashMap<String, Command>() {
+        @Override
+        public Command put(String key, Command value) {
+            // io.papermc.paper.command.brigadier.CommandBuilder.of(value.)
+            return super.put(key, value);
+        }
+
+        @Override
+        public Command get(Object key) {
+           // com.mojang.brigadier.tree.CommandNode<?> command = org.bukkit.Bukkit.getServer().getCommandDispatcher().findNode(List.of((String) key));
+            // Unsafe#getBukkit? Basically get it if it's stored of if it's some ki
+//            if (command instanceof io.papermc.paper.command.brigadier.PluginCommandNode pluginCommandNode) {
+//                return new ;
+//            } else if (command instanceof BukkitCommandNode bukkitCommandNode) {
+//                return bukkitCommandNode.getBukkitCommand();
+//            }
+
+            return super.get(key);
+        }
+    };
+    // Paper end
     private final Server server;
 
     public SimpleCommandMap(@NotNull final Server server) {
