From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Mon, 1 Aug 2022 22:50:29 -0400
Subject: [PATCH] Brigadier based command API


diff --git a/build.gradle.kts b/build.gradle.kts
index cad12a2632b9ebb569280441c42869685db1f31a..377e967eef17cf6407397245ea848b6bfc3ee73a 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -46,7 +46,7 @@ dependencies {
     implementation("org.ow2.asm:asm:9.4")
     implementation("org.ow2.asm:asm-commons:9.4")
     // Paper end
-
+    api("com.mojang:brigadier:1.0.18") // Paper, expose!
     api("org.apache.maven:maven-resolver-provider:3.8.5") // Paper, expose
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
     compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..1918953532ba9e10614510b7e4da66fb641c4af1
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java
@@ -0,0 +1,152 @@
+package io.papermc.paper.command.brigadier;
+
+import com.mojang.brigadier.CommandDispatcher;
+import com.mojang.brigadier.builder.ArgumentBuilder;
+import com.mojang.brigadier.tree.RootCommandNode;
+import org.bukkit.Bukkit;
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * A CommandBuilder gives you a proper way of registering brig commands.
+ * This will properly connect your brig commands to your plugin, ensuring that
+ * they are shown in /help, etc.
+ */
+@ApiStatus.Experimental
+public class CommandBuilder extends ArgumentBuilder<CommandSourceStack, CommandBuilder> {
+
+    private final Plugin plugin;
+    private final String literal;
+
+    private String description;
+    private List<String> aliases = List.of();
+
+    @ApiStatus.Internal
+    protected CommandBuilder(@NotNull Plugin plugin, @NotNull String literal) {
+        this.literal = literal;
+        this.plugin = plugin;
+    }
+
+    /**
+     * Creates a new command builder.
+     *
+     * @param plugin plugin
+     * @param name   command name
+     * @return builder
+     */
+    @NotNull
+    public static CommandBuilder of(@NotNull Plugin plugin, @NotNull String name) {
+        return new CommandBuilder(plugin, name);
+    }
+
+    /**
+     * Sets the aliases for this command.
+     *
+     * @param aliases alises
+     * @return self
+     */
+    @NotNull
+    public CommandBuilder aliases(@NotNull String... aliases) {
+        this.aliases = Arrays.asList(aliases);
+        return this.getThis();
+    }
+
+    /**
+     * Gets the description for this command.
+     *
+     * @return description
+     */
+    @NotNull
+    public String getDescription() {
+        return this.description;
+    }
+
+    /**
+     * Sets the description for this command.
+     *
+     * @param description description
+     * @return self
+     */
+    @NotNull
+    public CommandBuilder description(@NotNull String description) {
+        this.description = description;
+        return this.getThis();
+    }
+
+    /**
+     * Gets the plugin set for this command builder.
+     *
+     * @return self
+     */
+    @NotNull
+    public Plugin getPlugin() {
+        return this.plugin;
+    }
+
+    /**
+     * Returns self instance for builder pattern.
+     *
+     * @return self
+     */
+    @NotNull
+    @Override
+    protected CommandBuilder getThis() {
+        return this;
+    }
+
+    /**
+     * Builds the raw command node.
+     *
+     * @return raw node
+     */
+    @NotNull
+    @ApiStatus.Internal
+    @Override
+    public PluginCommandNode build() {
+        // See: LiteralArgumentBuilder
+        return new PluginCommandNode(this.literal, this);
+    }
+
+    /**
+     * Registers this command builder into the server
+     * dispatcher.
+     */
+    public void register() {
+        this.register(Bukkit.getServer().getCommandDispatcher());
+    }
+
+    /**
+     * Registers the command builder into the specified dispatcher.
+     *
+     * @param dispatcher dispatcher
+     */
+    public void register(@NotNull CommandDispatcher<CommandSourceStack> dispatcher) {
+        String identifier = this.plugin.getDescription().getName();
+        RootCommandNode<CommandSourceStack> dispatcherRoot = dispatcher.getRoot();
+        PluginCommandNode root = this.build();
+
+        dispatcherRoot.addChild(root);
+        // Add root redirect
+        this.registerRedirect(dispatcherRoot, root, identifier + ":" + this.literal);
+
+        // Add aliases
+        for (String alias : this.aliases) {
+            this.registerRedirect(dispatcherRoot, root, alias);
+            this.registerRedirect(dispatcherRoot, root, identifier + ":" + alias);
+        }
+    }
+
+    private void registerRedirect(@NotNull RootCommandNode<CommandSourceStack> dispatcherRoot, @NotNull PluginCommandNode node, @NotNull String name) {
+        PluginCommandNode aliasRoot = CommandBuilder.of(this.plugin, name)
+            .description(this.description)
+            .redirect(node)
+            .executes(node.getCommand())
+            .build();
+
+        dispatcherRoot.addChild(aliasRoot);
+    }
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java b/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java
new file mode 100644
index 0000000000000000000000000000000000000000..83651e20bf2bdf4e078dc2b8496f31917f6e8d34
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java
@@ -0,0 +1,47 @@
+package io.papermc.paper.command.brigadier;
+
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.command.CommandSender;
+import org.bukkit.entity.Entity;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Represents a vanilla command source stack which is used
+ * for command execution.
+ */
+@ApiStatus.NonExtendable
+@ApiStatus.Experimental
+public interface CommandSourceStack {
+
+    /**
+     * Gets the bukkit entity running this command.
+     * @return bukkit entity, null if not present
+     */
+    @Nullable
+    Entity getBukkitEntity();
+
+    /**
+     * Gets the world that this command is being executed in.
+     * @return world
+     */
+    @Nullable
+    World getBukkitWorld();
+
+    /**
+     * Gets the location that this command is being executed in.
+     * @return location
+     */
+    @Nullable
+    Location getBukkitLocation();
+
+    /**
+     * Gets the sender representing this source stack.
+     * @return source stack sender
+     */
+    @NotNull
+    CommandSender getBukkitSender();
+
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java
new file mode 100644
index 0000000000000000000000000000000000000000..94305f3f35d5f807381dcd3b26200f86cb0816e8
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java
@@ -0,0 +1,33 @@
+package io.papermc.paper.command.brigadier;
+
+import com.mojang.brigadier.Message;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.serializer.ComponentSerializer;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ServiceLoader;
+
+/**
+ * A component serializer that converts it to a brigadier compatible type.
+ */
+public interface MessageComponentSerializer extends ComponentSerializer<Component, Component, Message> {
+
+    /**
+     * A component serializer that converts it to a brigadier compatible type.
+     *
+     * @return serializer instance
+     */
+    @NotNull
+    static MessageComponentSerializer message() {
+        return Holder.PROVIDER;
+    }
+
+
+    @ApiStatus.Internal
+    class Holder {
+        public static final MessageComponentSerializer PROVIDER = ServiceLoader.load(MessageComponentSerializer.class)
+            .findFirst()
+            .orElseThrow();
+    }
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/PluginCommandNode.java b/src/main/java/io/papermc/paper/command/brigadier/PluginCommandNode.java
new file mode 100644
index 0000000000000000000000000000000000000000..ee073108df90ffdc58b524da03e76fa671757eeb
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/PluginCommandNode.java
@@ -0,0 +1,37 @@
+package io.papermc.paper.command.brigadier;
+
+import com.mojang.brigadier.tree.CommandNode;
+import com.mojang.brigadier.tree.LiteralCommandNode;
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+@ApiStatus.Internal
+public class PluginCommandNode extends LiteralCommandNode<CommandSourceStack> {
+
+    private final Plugin plugin;
+    private final String description;
+
+    public PluginCommandNode(@NotNull String literal, @NotNull CommandBuilder builder) {
+        super(
+            literal, builder.getCommand(), builder.getRequirement(),
+            builder.getRedirect(), builder.getRedirectModifier(), builder.isFork()
+        );
+        this.plugin = builder.getPlugin();
+        this.description = builder.getDescription();
+
+        for (CommandNode<CommandSourceStack> argument : builder.getArguments()) {
+            this.addChild(argument);
+        }
+    }
+
+    @NotNull
+    public Plugin getPlugin() {
+        return this.plugin;
+    }
+
+    @NotNull
+    public String getDescription() {
+        return this.description;
+    }
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/ArgumentResolver.java b/src/main/java/io/papermc/paper/command/brigadier/argument/ArgumentResolver.java
new file mode 100644
index 0000000000000000000000000000000000000000..14ff48fb772a9f0ca9a3d37f8a600a39d1d14a8c
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/argument/ArgumentResolver.java
@@ -0,0 +1,24 @@
+package io.papermc.paper.command.brigadier.argument;
+
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import io.papermc.paper.command.brigadier.CommandSourceStack;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Represents an argument which requires
+ * the context of a command source stack to be fully resolved.
+ * @param <T> type
+ */
+@ApiStatus.Experimental
+public interface ArgumentResolver<T> {
+
+    /**
+     * Resolves the argument with the given
+     * command source stack.
+     * @param sourceStack source stack
+     * @return resolved
+     */
+    @NotNull
+    T resolve(@NotNull CommandSourceStack sourceStack) throws CommandSyntaxException;
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/MessageArgumentResponse.java b/src/main/java/io/papermc/paper/command/brigadier/argument/MessageArgumentResponse.java
new file mode 100644
index 0000000000000000000000000000000000000000..8a39d092b9d54a7ab4f4e0f65f0c4b59db1e5db9
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/argument/MessageArgumentResponse.java
@@ -0,0 +1,41 @@
+package io.papermc.paper.command.brigadier.argument;
+
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import io.papermc.paper.command.brigadier.CommandSourceStack;
+import net.kyori.adventure.chat.SignedMessage;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Argument type returned from the Message argument.
+ */
+@ApiStatus.Experimental
+public interface MessageArgumentResponse {
+
+    /**
+     * Content of the message
+     *
+     * @return player made message
+     */
+    @NotNull
+    String content();
+
+    /**
+     * Resolves a signed message from this response.
+     * This will use context stored in the command source stack
+     * and signed arguments sent from the client.
+     * <p>
+     * In the case that signed message information isn't provided, a "system"
+     * signed message will be provided instead.
+     * @param argumentName argument name
+     * @param context context
+     * @return signed message
+     * @throws CommandSyntaxException syntax exception
+     */
+    @NotNull
+    CompletableFuture<SignedMessage> resolveSignedMessage(@NotNull String argumentName, @NotNull CommandContext<CommandSourceStack> context) throws CommandSyntaxException;
+
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArguments.java b/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArguments.java
new file mode 100644
index 0000000000000000000000000000000000000000..e43ce009c0245f008fa716e8f8bb6c13b8285cd0
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArguments.java
@@ -0,0 +1,190 @@
+package io.papermc.paper.command.brigadier.argument;
+
+import com.destroystokyo.paper.profile.PlayerProfile;
+import com.mojang.brigadier.arguments.ArgumentType;
+import io.papermc.paper.math.BlockPosition;
+import net.kyori.adventure.text.Component;
+import org.bukkit.Color;
+import org.bukkit.block.BlockState;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.ServiceLoader;
+import java.util.UUID;
+
+/**
+ * Vanilla argument registry
+ * <p>
+ * These provide rich argument parsing on the client and
+ * may also provide additional signing context.
+ */
+@ApiStatus.Experimental
+public final class VanillaArguments {
+
+    @NotNull
+    private static final VanillaArgumentProvider PROVIDER = ServiceLoader.load(VanillaArgumentProvider.class)
+        .findFirst()
+        .orElseThrow();
+
+    /**
+     * An ItemStack argument which provides rich parsing for
+     * specifying item material and item NBT information.
+     *
+     * @return argument
+     */
+    @NotNull
+    public static ArgumentType<ItemStack> itemStack() {
+        return PROVIDER.itemStack();
+    }
+
+    /**
+     * A blockstate argument which will provide rich parsing for specifying
+     * the specific block variant and then the block entity NBT if applicable.
+     *
+     * @return argument
+     */
+    @NotNull
+    public static ArgumentType<BlockState> blockState() {
+        return PROVIDER.blockState();
+    }
+
+    /**
+     * A uuid argument.
+     *
+     * @return argument
+     */
+    @NotNull
+    public static ArgumentType<UUID> uuid() {
+        return PROVIDER.uuid();
+    }
+
+    /**
+     * Represents a selector that can capture any
+     * entity.
+     *
+     * @return argument that takes one entity
+     */
+    @NotNull
+    public static ArgumentType<ArgumentResolver<Entity>> entity() {
+        return PROVIDER.entity();
+    }
+
+    /**
+     * Represents a selector that can capture multiple
+     * entities.
+     *
+     * @return argument that takes multiple entities
+     */
+    @NotNull
+    public static ArgumentType<ArgumentResolver<Collection<Entity>>> entities() {
+        return PROVIDER.entities();
+    }
+
+    /**
+     * Represents a selector that can capture a
+     * player entity.
+     *
+     * @return argument that takes one player
+     */
+    @NotNull
+    public static ArgumentType<ArgumentResolver<Player>> player() {
+        return PROVIDER.player();
+    }
+
+    /**
+     * Represents a selector that can capture multiple
+     * player entities.
+     *
+     * @return argument that takes multiple players
+     */
+    @NotNull
+    public static ArgumentType<ArgumentResolver<Collection<Player>>> players() {
+        return PROVIDER.players();
+    }
+
+    /**
+     * Represents a selector that provides list
+     * of player profiles
+     *
+     * @return player profile arguments
+     */
+    @NotNull
+    public static ArgumentType<ArgumentResolver<Collection<PlayerProfile>>> playerProfiles() {
+        return PROVIDER.playerProfiles();
+    }
+
+    /**
+     * A block position argument.
+     *
+     * @return argument
+     */
+    @NotNull
+    public static ArgumentType<ArgumentResolver<BlockPosition>> blockPos() {
+        return PROVIDER.blockPos();
+    }
+
+    /**
+     * A signed message argument.
+     * This argument can be resolved to retreieve the underlying
+     * signed message.
+     *
+     * @return argument
+     */
+    @NotNull
+    public static ArgumentType<MessageArgumentResponse> signedMessage() {
+        return PROVIDER.signedMessage();
+    }
+
+
+    /**
+     * A component argument.
+     *
+     * @return argument
+     */
+    @NotNull
+    public static ArgumentType<Component> component() {
+        return PROVIDER.component();
+    }
+
+    @ApiStatus.Internal
+    interface VanillaArgumentProvider {
+
+        @NotNull
+        ArgumentType<ItemStack> itemStack();
+
+        @NotNull
+        ArgumentType<BlockState> blockState();
+
+        @NotNull
+        ArgumentType<MessageArgumentResponse> signedMessage();
+
+        @NotNull
+        ArgumentType<UUID> uuid();
+
+        @NotNull
+        ArgumentType<ArgumentResolver<Entity>> entity();
+
+        @NotNull
+        ArgumentType<ArgumentResolver<Player>> player();
+
+        @NotNull
+        ArgumentType<ArgumentResolver<Collection<Entity>>> entities();
+
+        @NotNull
+        ArgumentType<ArgumentResolver<Collection<Player>>> players();
+
+        @NotNull
+        ArgumentType<ArgumentResolver<Collection<PlayerProfile>>> playerProfiles();
+
+        @NotNull
+        ArgumentType<ArgumentResolver<BlockPosition>> blockPos();
+
+        @NotNull
+        ArgumentType<Component> component();
+    }
+
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/WrapperArgumentType.java b/src/main/java/io/papermc/paper/command/brigadier/argument/WrapperArgumentType.java
new file mode 100644
index 0000000000000000000000000000000000000000..c5c912e792297ae2129e9ff56c7bcad82c0493ac
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/argument/WrapperArgumentType.java
@@ -0,0 +1,86 @@
+package io.papermc.paper.command.brigadier.argument;
+
+import com.mojang.brigadier.StringReader;
+import com.mojang.brigadier.arguments.ArgumentType;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+
+/**
+ * An argument type that wraps around another argument.
+ * This argument is special in that the argument type that is wrapped will
+ * be sent to the client, not the wrapper itself.
+ *
+ * @param <T> type
+ * @param <B> base wrapping type
+ */
+@ApiStatus.Experimental
+public abstract class WrapperArgumentType<T, B> implements ArgumentType<T> {
+
+    @NotNull
+    private final ArgumentType<B> baseArgumentType;
+
+    public WrapperArgumentType(@NotNull ArgumentType<B> baseArgumentType) {
+        this.baseArgumentType = baseArgumentType;
+    }
+
+    /**
+     * Parses the argument using the wrapped argument type then converting
+     * it.
+     * @param reader string reader
+     * @return value
+     * @throws CommandSyntaxException if an error occurs while parsing
+     */
+    @NotNull
+    @Override
+    public T parse(@NotNull StringReader reader) throws CommandSyntaxException {
+        return this.convert(this.baseArgumentType.parse(reader));
+    }
+
+    /**
+     * Converts the type from the wrapped argument to this argument.
+     *
+     * @param baseType wrapped argument provided value
+     * @return converted value
+     * @throws CommandSyntaxException if an exception occurs while parsing
+     */
+    @NotNull
+    public abstract T convert(@NotNull B baseType) throws CommandSyntaxException;
+
+    /**
+     * Gets the wrapped argument.
+     *
+     * @return wrapped
+     */
+    @NotNull
+    public ArgumentType<B> getWrapped() {
+        return this.baseArgumentType;
+    }
+
+    /**
+     * Cannot be controlled by the server.
+     * Returned in cases where there are multiple arguments in the same node.
+     * This helps differentiate and tell the player what the possible inputs are.
+     *
+     * @return client set examples
+     */
+    @NotNull
+    @Override
+    public final Collection<String> getExamples() {
+        return this.baseArgumentType.getExamples();
+    }
+
+    /**
+     * Returns if the argument itself should handle sugggestion
+     * parsing instead of the wrapper.
+     *
+     * @return handles suggestions
+     */
+    public boolean handleSuggestions() {
+        return false;
+    }
+
+
+}
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index f1af93319d6b29d7bc709e7e273e8abe26c7b4af..04198d13c59bce7b78ebb30fa2439f64fa2e164b 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2335,6 +2335,7 @@ public final class Bukkit {
      * @return the default message
      */
     @NotNull
+    @Deprecated // Used only for bukkit commands
     public static net.kyori.adventure.text.Component permissionMessage() {
         return server.permissionMessage();
     }
@@ -2455,6 +2456,11 @@ public final class Bukkit {
     public static @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer() {
         return server.getPotionBrewer();
     }
+
+    @NotNull
+    public static com.mojang.brigadier.CommandDispatcher<io.papermc.paper.command.brigadier.CommandSourceStack> getCommandDispatcher() {
+        return server.getCommandDispatcher();
+    }
     // Paper end
 
     @NotNull
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 8621dba7f10ab822b5b99ce0d05da58823605cb6..22374a34f0592ca7c06080c81b7dc1cec1139d34 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2030,6 +2030,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      *
      * @return the default message
      */
+    @Deprecated // Paper - only used for bukkit commands
     @NotNull net.kyori.adventure.text.Component permissionMessage();
 
     /**
@@ -2135,5 +2136,8 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * @return the potion brewer
      */
     @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer();
+
+    @NotNull
+    com.mojang.brigadier.CommandDispatcher<io.papermc.paper.command.brigadier.CommandSourceStack> getCommandDispatcher();
     // Paper end
 }
diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index b92255a9c87620f46adb140689b1cd328a476d61..50807b2718676335d981aaca6697718fce49781f 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -244,5 +244,7 @@ public interface UnsafeValues {
      * @throws IllegalStateException if no biome by the given key is registered.
      */
     void setBiomeKey(RegionAccessor accessor, int x, int y, int z, NamespacedKey biomeKey);
+
+    java.util.Map<String, org.bukkit.command.Command> getCommandMirrorMap();
     // Paper end
 }
diff --git a/src/main/java/org/bukkit/command/Command.java b/src/main/java/org/bukkit/command/Command.java
index b791358f90fe92bc2264d9a26492245763813af3..28fc568a033f0c8c58292983b5b5f2adec1987c9 100644
--- a/src/main/java/org/bukkit/command/Command.java
+++ b/src/main/java/org/bukkit/command/Command.java
@@ -488,4 +488,12 @@ public abstract class Command {
     public String toString() {
         return getClass().getName() + '(' + name + ')';
     }
+
+    // Paper start
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public boolean canBeOverriden() { return false; }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public String registeredNamespace;
+    // Paper end
 }
diff --git a/src/main/java/org/bukkit/command/CommandMap.java b/src/main/java/org/bukkit/command/CommandMap.java
index 864c263bbd4dd6dd7c37a74b39b1a40a884d0731..88d92350b59333c2148190c5e4974f254b0139a9 100644
--- a/src/main/java/org/bukkit/command/CommandMap.java
+++ b/src/main/java/org/bukkit/command/CommandMap.java
@@ -5,6 +5,7 @@ import org.bukkit.Location;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+@Deprecated(forRemoval = true) // Paper
 public interface CommandMap {
 
     /**
diff --git a/src/main/java/org/bukkit/command/FormattedCommandAlias.java b/src/main/java/org/bukkit/command/FormattedCommandAlias.java
index 9d4f553c04784cca63901a56a7aea62a5cae1d72..abe256e1e45ce28036da4aa1586715bc8a1a3414 100644
--- a/src/main/java/org/bukkit/command/FormattedCommandAlias.java
+++ b/src/main/java/org/bukkit/command/FormattedCommandAlias.java
@@ -117,7 +117,7 @@ public class FormattedCommandAlias extends Command {
             index = formatString.indexOf('$', index);
         }
 
-        return formatString;
+        return formatString.trim(); // Paper - Causes an extra space at the end, breaks with brig commands
     }
 
     @NotNull
diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index ac9a28922f8a556944a4c3649d74c32c622f0cb0..92bab886dde8d82e3fe53b79c7e0189b5172e040 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -21,8 +21,9 @@ import org.bukkit.util.StringUtil;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+@Deprecated // Paper
 public class SimpleCommandMap implements CommandMap {
-    protected final Map<String, Command> knownCommands = new HashMap<String, Command>();
+    protected final Map<String, Command> knownCommands = org.bukkit.Bukkit.getUnsafe().getCommandMirrorMap(); // Paper
     private final Server server;
 
     public SimpleCommandMap(@NotNull final Server server) {
@@ -101,6 +102,7 @@ public class SimpleCommandMap implements CommandMap {
      * @return true if command was registered, false otherwise.
      */
     private synchronized boolean register(@NotNull String label, @NotNull Command command, boolean isAlias, @NotNull String fallbackPrefix) {
+        command.registeredNamespace = fallbackPrefix; // Paper
         knownCommands.put(fallbackPrefix + ":" + label, command);
         if ((command instanceof BukkitCommand || isAlias) && knownCommands.containsKey(label)) {
             // Request is for an alias/fallback command and it conflicts with
@@ -114,7 +116,9 @@ public class SimpleCommandMap implements CommandMap {
         // If the command exists but is an alias we overwrite it, otherwise we return
         Command conflict = knownCommands.get(label);
         if (conflict != null && conflict.getLabel().equals(label)) {
+            if (!conflict.canBeOverriden()) { // Paper
             return false;
+            } // Paper
         }
 
         if (!isAlias) {
