From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Mon, 1 Aug 2022 22:50:29 -0400
Subject: [PATCH] Brigadier based command API


diff --git a/build.gradle.kts b/build.gradle.kts
index 149f9088fe806467656e8b1c4157df60fda69ba7..24b533f277092d1e70ca1d5e5e8f67bd1e3da8fd 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -51,6 +51,7 @@ dependencies {
     implementation("org.ow2.asm:asm-commons:9.4")
     // Paper end
 
+    api("com.mojang:brigadier:1.0.18") // Paper, expose!
     api("org.apache.maven:maven-resolver-provider:3.8.5")
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
     compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..5f07c0c8114ebe3c363567451768111faa38fdb5
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java
@@ -0,0 +1,76 @@
+package io.papermc.paper.command.brigadier;
+
+import com.mojang.brigadier.builder.ArgumentBuilder;
+import com.mojang.brigadier.tree.CommandNode;
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ServiceLoader;
+
+@ApiStatus.Experimental
+@ApiStatus.NonExtendable
+public abstract class CommandBuilder extends ArgumentBuilder<CommandSourceStack, CommandBuilder> {
+
+    private static final CommandBuilderProvider PROVIDER = ServiceLoader.load(CommandBuilderProvider.class)
+        .findFirst()
+        .orElseThrow();
+
+    /**
+     * Creates a new command builder.
+     *
+     * @param plugin plugin
+     * @param name   command name
+     * @return builder
+     */
+    @NotNull
+    public static CommandBuilder newCommandBuilder(@NotNull Plugin plugin, @NotNull String name) {
+        return PROVIDER.create(plugin, name);
+    }
+
+    /**
+     * Sets the aliases for this command.
+     *
+     * @param aliases alises
+     * @return self
+     */
+    @NotNull
+    public abstract CommandBuilder aliases(@NotNull String... aliases);
+
+    /**
+     * Gets the description for this command.
+     *
+     * @return description
+     */
+    @Nullable
+    public abstract String getDescription();
+
+    /**
+     * Sets the description for this command.
+     *
+     * @param description description
+     * @return self
+     */
+    @NotNull
+    public abstract CommandBuilder description(@Nullable String description);
+
+    /**
+     * Gets the plugin set for this command builder.
+     *
+     * @return self
+     */
+    @NotNull
+    public abstract Plugin getPlugin();
+
+    /**
+     * Builds the underlying command argument node directly.
+     * @return command argument node
+     */
+    @Deprecated
+    @Contract("-> fail")
+    @Override
+    public abstract CommandNode<CommandSourceStack> build();
+
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandBuilderProvider.java b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilderProvider.java
new file mode 100644
index 0000000000000000000000000000000000000000..3473c29dfe411377bff4bd005eba7b130024dd4a
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilderProvider.java
@@ -0,0 +1,9 @@
+package io.papermc.paper.command.brigadier;
+
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Internal
+interface CommandBuilderProvider {
+    CommandBuilder create(Plugin plugin, String name);
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java b/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java
new file mode 100644
index 0000000000000000000000000000000000000000..deb6ccfb5b0456280b99325e839cd72fca50a86d
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java
@@ -0,0 +1,58 @@
+package io.papermc.paper.command.brigadier;
+
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import io.papermc.paper.command.brigadier.argument.ArgumentResolver;
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.command.CommandSender;
+import org.bukkit.entity.Entity;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Represents a vanilla command source stack which is used
+ * for command execution.
+ */
+@ApiStatus.NonExtendable
+@ApiStatus.Experimental
+public interface CommandSourceStack {
+
+    /**
+     * Gets the location that this command is being executed at.
+     * @return location
+     */
+    @NotNull
+    Location getLocation();
+
+    /**
+     * Gets the source of from where this command was executed from
+     * @return source type
+     */
+    @NotNull
+    CommandSender getSender();
+
+    /**
+     * Gets the entity that triggered the execution of this command.
+     * May not always be the command source.
+     */
+    @Nullable
+    Entity getExecutor();
+
+    /**
+     * Resolves an argument in this command source stack.
+     *
+     * @param context command context
+     * @param name name of argument
+     * @return resolved argument
+     * @param <V> resolved argument type
+     */
+    @NotNull
+    @SuppressWarnings("unchecked")
+    default <V> V getResolvedArgument(@NotNull CommandContext<CommandSourceStack> context, @NotNull final String name) throws CommandSyntaxException {
+        ArgumentResolver<?> resolver = context.getArgument(name, ArgumentResolver.class);
+        return (V) resolver.resolve(this);
+    }
+
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/Commands.java b/src/main/java/io/papermc/paper/command/brigadier/Commands.java
new file mode 100644
index 0000000000000000000000000000000000000000..2623aedc1e0d4e7e849167f95b2be4e4cb7dde0b
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/Commands.java
@@ -0,0 +1,88 @@
+package io.papermc.paper.command.brigadier;
+
+import com.mojang.brigadier.CommandDispatcher;
+import com.mojang.brigadier.arguments.ArgumentType;
+import com.mojang.brigadier.builder.LiteralArgumentBuilder;
+import com.mojang.brigadier.builder.RequiredArgumentBuilder;
+import org.bukkit.command.Command;
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Represents the general holder for commands.
+ */
+@ApiStatus.Experimental
+public interface Commands {
+
+    /**
+     * Static constructor utility to create a literal command node builder.
+     * @param literal literal name
+     * @return builder
+     */
+    @NotNull
+    static LiteralArgumentBuilder<CommandSourceStack> literal(@NotNull String literal) {
+        return LiteralArgumentBuilder.literal(literal);
+    }
+
+    /**
+     * Static constructor utility to create a required argument builder.
+     *
+     * @param name name
+     * @param argumentType type
+     * @return value
+     * @param <T> arg type
+     */
+    @NotNull
+    static <T> RequiredArgumentBuilder<CommandSourceStack, T> argument(@NotNull String name, @NotNull ArgumentType<T> argumentType) {
+        return RequiredArgumentBuilder.argument(name, argumentType);
+    }
+
+    /**
+     * Gets the underlying {@link CommandDispatcher}.
+     *
+     * <p><b>Note:</b> This is a delicate API that must be used with care to ensure a consistent user experience.</p>
+     *
+     * <p>When registering commands, it should be preferred to use {@link #register(CommandBuilder)}
+     * over directly registering to the dispatcher wherever possible. {@link CommandBuilder CommandBuilders}
+     * automatically handle command namespacing, command help, plugin association with commands, and more.</p>
+     *
+     * <p>Example use cases for this method <b>may</b> include:
+     * <ul>
+     *   <li>Implementing integration between an external command framework and Paper (although {@link CommandBuilder} should still be preferred where possible)</li>
+     *   <li>Registering new child nodes to an existing plugin command (for example an "addon" plugin to another plugin may want to do this)</li>
+     *   <li>Retrieving existing command nodes to build redirects</li>
+     * </ul></p>
+     * @return dispatcher
+     */
+    @NotNull
+    CommandDispatcher<CommandSourceStack> getDispatcher();
+
+    /**
+     * Registers a {@link CommandBuilder} into the underlying
+     * {@link CommandDispatcher}.
+     * <p>Commands have certain the overriding behavior of:
+     * <ul>
+     *   <li>Aliases will not override already existing commands (excluding namespaced ones)</li>
+     *   <li>The main command/namespaced label will override already existing commands</li>
+     * </ul></p>
+     * @param builder command builder
+     * @return true if at least one label was properly registered
+     * into the dispatcher
+     */
+    boolean register(@NotNull CommandBuilder builder);
+
+    /**
+     * Registers a command under the same logic as {@link Commands#register(CommandBuilder)}.
+     *
+     * @param plugin plugin instance
+     * @param label command label
+     * @param description comman description, as shown in /help
+     * @param simpleCommand command implementation
+     * @return true if at least one label was properly registered
+     * into the dispatcher
+     */
+    boolean register(@NotNull Plugin plugin, @NotNull String label, @NotNull String description, @NotNull SimpleCommand simpleCommand);
+
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java
new file mode 100644
index 0000000000000000000000000000000000000000..4cbd2e3f79a0a550d8b1a249996b2058b24717a9
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java
@@ -0,0 +1,26 @@
+package io.papermc.paper.command.brigadier;
+
+import com.mojang.brigadier.Message;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.serializer.ComponentSerializer;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ServiceLoader;
+
+/**
+ * A component serializer that converts it to a brigadier compatible type.
+ */
+@ApiStatus.Experimental
+public interface MessageComponentSerializer extends ComponentSerializer<Component, Component, Message> {
+
+    /**
+     * A component serializer that converts it to a brigadier compatible type.
+     *
+     * @return serializer instance
+     */
+    @NotNull
+    static MessageComponentSerializer message() {
+        return MessageComponentSerializerHolder.PROVIDER;
+    }
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerHolder.java b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerHolder.java
new file mode 100644
index 0000000000000000000000000000000000000000..f3b8cc535cfa74a16cd35ed5677c6958b894046d
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerHolder.java
@@ -0,0 +1,10 @@
+package io.papermc.paper.command.brigadier;
+
+import java.util.ServiceLoader;
+
+final class MessageComponentSerializerHolder {
+
+    public static final MessageComponentSerializer PROVIDER = ServiceLoader.load(MessageComponentSerializer.class)
+        .findFirst()
+        .orElseThrow();
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/SimpleCommand.java b/src/main/java/io/papermc/paper/command/brigadier/SimpleCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..8e8843dd8e27d98f2d09afdd8c1f1051b151564b
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/SimpleCommand.java
@@ -0,0 +1,37 @@
+package io.papermc.paper.command.brigadier;
+
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.suggestion.Suggestions;
+import com.mojang.brigadier.suggestion.SuggestionsBuilder;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * This represents a simple command implementation that is wrapped
+ * around brigadier.
+ */
+@ApiStatus.Experimental
+public interface SimpleCommand {
+
+    /**
+     * Executes the command with the given context and arguments.
+     *
+     * @param context The context of the command.
+     * @param args The arguments of the command.
+     * @return The result of the command execution.
+     */
+    int execute(CommandContext<CommandSourceStack> context, String[] args);
+
+    /**
+     * Suggests possible completions for the given command context and arguments.
+     *
+     * @param context The context of the command.
+     * @param builder The suggestions builder.
+     * @param args The arguments of the command.
+     * @return A completable future containing the suggestions.
+     */
+    @NotNull
+    CompletableFuture<Suggestions> suggest(CommandContext<CommandSourceStack> context, SuggestionsBuilder builder, String[] args);
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/ArgumentResolver.java b/src/main/java/io/papermc/paper/command/brigadier/argument/ArgumentResolver.java
new file mode 100644
index 0000000000000000000000000000000000000000..14ff48fb772a9f0ca9a3d37f8a600a39d1d14a8c
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/argument/ArgumentResolver.java
@@ -0,0 +1,24 @@
+package io.papermc.paper.command.brigadier.argument;
+
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import io.papermc.paper.command.brigadier.CommandSourceStack;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Represents an argument which requires
+ * the context of a command source stack to be fully resolved.
+ * @param <T> type
+ */
+@ApiStatus.Experimental
+public interface ArgumentResolver<T> {
+
+    /**
+     * Resolves the argument with the given
+     * command source stack.
+     * @param sourceStack source stack
+     * @return resolved
+     */
+    @NotNull
+    T resolve(@NotNull CommandSourceStack sourceStack) throws CommandSyntaxException;
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/CustomArgumentType.java b/src/main/java/io/papermc/paper/command/brigadier/argument/CustomArgumentType.java
new file mode 100644
index 0000000000000000000000000000000000000000..50a54e913a2aad1dcf117814de353e9b6fcadcc6
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/argument/CustomArgumentType.java
@@ -0,0 +1,87 @@
+package io.papermc.paper.command.brigadier.argument;
+
+import com.mojang.brigadier.StringReader;
+import com.mojang.brigadier.arguments.ArgumentType;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+
+/**
+ * An argument type that wraps around another argument.
+ * This argument is special in that the argument type that is wrapped will
+ * be sent to the client, not the wrapper itself.
+ *
+ * @param <T> type
+ * @param <B> base wrapping type
+ */
+@ApiStatus.Experimental
+public abstract class CustomArgumentType<T, B> implements ArgumentType<T> {
+
+    @NotNull
+    private final ArgumentType<B> vanillaPrimitiveType;
+
+    public CustomArgumentType(@NotNull ArgumentType<B> primitive) {
+        VanillaArguments.PROVIDER.validatePrimitiveType(primitive);
+        this.vanillaPrimitiveType = primitive;
+    }
+
+    /**
+     * Parses the argument using the wrapped argument type then converting
+     * it.
+     * @param reader string reader
+     * @return value
+     * @throws CommandSyntaxException if an error occurs while parsing
+     */
+    @NotNull
+    @Override
+    public T parse(@NotNull StringReader reader) throws CommandSyntaxException {
+        return this.convert(this.vanillaPrimitiveType.parse(reader));
+    }
+
+    /**
+     * Converts the type from the wrapped argument to this argument.
+     *
+     * @param baseType wrapped argument provided value
+     * @return converted value
+     * @throws CommandSyntaxException if an exception occurs while parsing
+     */
+    @NotNull
+    public abstract T convert(@NotNull B baseType) throws CommandSyntaxException;
+
+    /**
+     * Gets the primitive type that this argument uses,
+     * or the type that is sent to the client.
+     *
+     * @return primitive argument
+     */
+    @NotNull
+    public final ArgumentType<B> getPrimitiveType() {
+        return this.vanillaPrimitiveType;
+    }
+
+    /**
+     * Cannot be controlled by the server.
+     * Returned in cases where there are multiple arguments in the same node.
+     * This helps differentiate and tell the player what the possible inputs are.
+     *
+     * @return client set examples
+     */
+    @NotNull
+    @Override
+    public final Collection<String> getExamples() {
+        return this.vanillaPrimitiveType.getExamples();
+    }
+
+    /**
+     * Returns if the argument itself should handle sugggestion
+     * parsing instead of the wrapper.
+     *
+     * @return handles suggestions
+     */
+    public boolean handleSuggestions() {
+        return false;
+    }
+
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/MessageArgumentResponse.java b/src/main/java/io/papermc/paper/command/brigadier/argument/MessageArgumentResponse.java
new file mode 100644
index 0000000000000000000000000000000000000000..8a39d092b9d54a7ab4f4e0f65f0c4b59db1e5db9
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/argument/MessageArgumentResponse.java
@@ -0,0 +1,41 @@
+package io.papermc.paper.command.brigadier.argument;
+
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import io.papermc.paper.command.brigadier.CommandSourceStack;
+import net.kyori.adventure.chat.SignedMessage;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Argument type returned from the Message argument.
+ */
+@ApiStatus.Experimental
+public interface MessageArgumentResponse {
+
+    /**
+     * Content of the message
+     *
+     * @return player made message
+     */
+    @NotNull
+    String content();
+
+    /**
+     * Resolves a signed message from this response.
+     * This will use context stored in the command source stack
+     * and signed arguments sent from the client.
+     * <p>
+     * In the case that signed message information isn't provided, a "system"
+     * signed message will be provided instead.
+     * @param argumentName argument name
+     * @param context context
+     * @return signed message
+     * @throws CommandSyntaxException syntax exception
+     */
+    @NotNull
+    CompletableFuture<SignedMessage> resolveSignedMessage(@NotNull String argumentName, @NotNull CommandContext<CommandSourceStack> context) throws CommandSyntaxException;
+
+}
diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArguments.java b/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArguments.java
new file mode 100644
index 0000000000000000000000000000000000000000..5da91abdee4d78a949e78c878f6b036df9374b14
--- /dev/null
+++ b/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArguments.java
@@ -0,0 +1,192 @@
+package io.papermc.paper.command.brigadier.argument;
+
+import com.destroystokyo.paper.profile.PlayerProfile;
+import com.mojang.brigadier.arguments.ArgumentType;
+import io.papermc.paper.math.BlockPosition;
+import net.kyori.adventure.text.Component;
+import org.bukkit.Color;
+import org.bukkit.block.BlockState;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.ServiceLoader;
+import java.util.UUID;
+
+/**
+ * Vanilla argument registry
+ * <p>
+ * These provide rich argument parsing on the client and
+ * may also provide additional signing context.
+ */
+@ApiStatus.Experimental
+public final class VanillaArguments {
+
+    @NotNull
+    static final VanillaArgumentProvider PROVIDER = ServiceLoader.load(VanillaArgumentProvider.class)
+        .findFirst()
+        .orElseThrow();
+
+    /**
+     * An ItemStack argument which provides rich parsing for
+     * specifying item material and item NBT information.
+     *
+     * @return argument
+     */
+    @NotNull
+    public static ArgumentType<ItemStack> itemStack() {
+        return PROVIDER.itemStack();
+    }
+
+    /**
+     * A blockstate argument which will provide rich parsing for specifying
+     * the specific block variant and then the block entity NBT if applicable.
+     *
+     * @return argument
+     */
+    @NotNull
+    public static ArgumentType<BlockState> blockState() {
+        return PROVIDER.blockState();
+    }
+
+    /**
+     * A uuid argument.
+     *
+     * @return argument
+     */
+    @NotNull
+    public static ArgumentType<UUID> uuid() {
+        return PROVIDER.uuid();
+    }
+
+    /**
+     * Represents a selector that can capture any
+     * entity.
+     *
+     * @return argument that takes one entity
+     */
+    @NotNull
+    public static ArgumentType<ArgumentResolver<Entity>> entity() {
+        return PROVIDER.entity();
+    }
+
+    /**
+     * Represents a selector that can capture multiple
+     * entities.
+     *
+     * @return argument that takes multiple entities
+     */
+    @NotNull
+    public static ArgumentType<ArgumentResolver<Collection<Entity>>> entities() {
+        return PROVIDER.entities();
+    }
+
+    /**
+     * Represents a selector that can capture a
+     * player entity.
+     *
+     * @return argument that takes one player
+     */
+    @NotNull
+    public static ArgumentType<ArgumentResolver<Player>> player() {
+        return PROVIDER.player();
+    }
+
+    /**
+     * Represents a selector that can capture multiple
+     * player entities.
+     *
+     * @return argument that takes multiple players
+     */
+    @NotNull
+    public static ArgumentType<ArgumentResolver<Collection<Player>>> players() {
+        return PROVIDER.players();
+    }
+
+    /**
+     * Represents a selector that provides list
+     * of player profiles
+     *
+     * @return player profile arguments
+     */
+    @NotNull
+    public static ArgumentType<ArgumentResolver<Collection<PlayerProfile>>> playerProfiles() {
+        return PROVIDER.playerProfiles();
+    }
+
+    /**
+     * A block position argument.
+     *
+     * @return argument
+     */
+    @NotNull
+    public static ArgumentType<ArgumentResolver<BlockPosition>> blockPos() {
+        return PROVIDER.blockPos();
+    }
+
+    /**
+     * A signed message argument.
+     * This argument can be resolved to retreieve the underlying
+     * signed message.
+     *
+     * @return argument
+     */
+    @NotNull
+    public static ArgumentType<MessageArgumentResponse> signedMessage() {
+        return PROVIDER.signedMessage();
+    }
+
+
+    /**
+     * A component argument.
+     *
+     * @return argument
+     */
+    @NotNull
+    public static ArgumentType<Component> component() {
+        return PROVIDER.component();
+    }
+
+    @ApiStatus.Internal
+    interface VanillaArgumentProvider {
+
+        @NotNull
+        ArgumentType<ItemStack> itemStack();
+
+        @NotNull
+        ArgumentType<BlockState> blockState();
+
+        @NotNull
+        ArgumentType<MessageArgumentResponse> signedMessage();
+
+        @NotNull
+        ArgumentType<UUID> uuid();
+
+        @NotNull
+        ArgumentType<ArgumentResolver<Entity>> entity();
+
+        @NotNull
+        ArgumentType<ArgumentResolver<Player>> player();
+
+        @NotNull
+        ArgumentType<ArgumentResolver<Collection<Entity>>> entities();
+
+        @NotNull
+        ArgumentType<ArgumentResolver<Collection<Player>>> players();
+
+        @NotNull
+        ArgumentType<ArgumentResolver<Collection<PlayerProfile>>> playerProfiles();
+
+        @NotNull
+        ArgumentType<ArgumentResolver<BlockPosition>> blockPos();
+
+        @NotNull
+        ArgumentType<Component> component();
+
+        void validatePrimitiveType(@NotNull ArgumentType<?> type);
+    }
+
+}
diff --git a/src/main/java/io/papermc/paper/event/server/ServerResourcesLoadEvent.java b/src/main/java/io/papermc/paper/event/server/ServerResourcesLoadEvent.java
index 2f468510309ecaa8c1b8d8d96edaa8bbd55111d0..43dfa48bfd77e3d9350cbc52b78300bcff97a113 100644
--- a/src/main/java/io/papermc/paper/event/server/ServerResourcesLoadEvent.java
+++ b/src/main/java/io/papermc/paper/event/server/ServerResourcesLoadEvent.java
@@ -1,5 +1,8 @@
 package io.papermc.paper.event.server;
 
+import com.mojang.brigadier.CommandDispatcher;
+import io.papermc.paper.command.brigadier.CommandSourceStack;
+import io.papermc.paper.command.brigadier.Commands;
 import org.bukkit.event.HandlerList;
 import org.bukkit.event.server.ServerEvent;
 import org.jetbrains.annotations.NotNull;
@@ -9,13 +12,22 @@ import org.jetbrains.annotations.NotNull;
  * <p>
  *     Intended for use to register custom recipes, advancements, and commands.
  *     This is executed on reload as well.
- *
  * </p>
  */
 public class ServerResourcesLoadEvent extends ServerEvent {
 
     public static final HandlerList HANDLER_LIST = new HandlerList();
 
+    private final Commands commands;
+
+    public ServerResourcesLoadEvent(Commands commands) {
+        this.commands = commands;
+    }
+
+    public Commands getCommands() {
+        return commands;
+    }
+
     @NotNull
     public static HandlerList getHandlerList() {
         return HANDLER_LIST;
diff --git a/src/main/java/io/papermc/paper/event/server/ServerResourcesReloadedEvent.java b/src/main/java/io/papermc/paper/event/server/ServerResourcesReloadedEvent.java
index 16f3969dca58746f93853680397849fefde099e0..8984cc89d88555b4ced701541cfa667becaca9cc 100644
--- a/src/main/java/io/papermc/paper/event/server/ServerResourcesReloadedEvent.java
+++ b/src/main/java/io/papermc/paper/event/server/ServerResourcesReloadedEvent.java
@@ -1,5 +1,6 @@
 package io.papermc.paper.event.server;
 
+import io.papermc.paper.command.brigadier.Commands;
 import org.jetbrains.annotations.NotNull;
 
 /**
@@ -12,7 +13,8 @@ public class ServerResourcesReloadedEvent extends ServerResourcesLoadEvent {
 
     private final Cause cause;
 
-    public ServerResourcesReloadedEvent(@NotNull Cause cause) {
+    public ServerResourcesReloadedEvent(@NotNull Cause cause, Commands dispatcher) {
+        super(dispatcher);
         this.cause = cause;
     }
 
diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index c661eab343ae76488de701630424e2d589f44fc0..c922111e231e61b889252aae3adebda0a4d698b4 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -249,5 +249,7 @@ public interface UnsafeValues {
      * @throws IllegalStateException if no biome by the given key is registered.
      */
     void setBiomeKey(RegionAccessor accessor, int x, int y, int z, NamespacedKey biomeKey);
+
+    java.util.Map<String, org.bukkit.command.Command> getCommandMirrorMap();
     // Paper end
 }
diff --git a/src/main/java/org/bukkit/command/Command.java b/src/main/java/org/bukkit/command/Command.java
index b791358f90fe92bc2264d9a26492245763813af3..d816d1677dface8cb3035d5adb9919aa75bd8e44 100644
--- a/src/main/java/org/bukkit/command/Command.java
+++ b/src/main/java/org/bukkit/command/Command.java
@@ -488,4 +488,9 @@ public abstract class Command {
     public String toString() {
         return getClass().getName() + '(' + name + ')';
     }
+
+    // Paper start
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public boolean canBeOverriden() { return false; }
+    // Paper end
 }
diff --git a/src/main/java/org/bukkit/command/FormattedCommandAlias.java b/src/main/java/org/bukkit/command/FormattedCommandAlias.java
index 9d4f553c04784cca63901a56a7aea62a5cae1d72..abe256e1e45ce28036da4aa1586715bc8a1a3414 100644
--- a/src/main/java/org/bukkit/command/FormattedCommandAlias.java
+++ b/src/main/java/org/bukkit/command/FormattedCommandAlias.java
@@ -117,7 +117,7 @@ public class FormattedCommandAlias extends Command {
             index = formatString.indexOf('$', index);
         }
 
-        return formatString;
+        return formatString.trim(); // Paper - Causes an extra space at the end, breaks with brig commands
     }
 
     @NotNull
diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index ac9a28922f8a556944a4c3649d74c32c622f0cb0..29a6227e51588c2df427d3b5c5a88189da361216 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -22,10 +22,18 @@ import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 public class SimpleCommandMap implements CommandMap {
-    protected final Map<String, Command> knownCommands = new HashMap<String, Command>();
+    protected final Map<String, Command> knownCommands; // Paper
     private final Server server;
 
     public SimpleCommandMap(@NotNull final Server server) {
+    // Paper start
+        this(server, org.bukkit.Bukkit.getUnsafe().getCommandMirrorMap());
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public SimpleCommandMap(@NotNull final Server server, Map<String, Command> backing) {
+        this.knownCommands = backing;
+    // Paper end
         this.server = server;
         setDefaultCommands();
     }
@@ -102,7 +110,10 @@ public class SimpleCommandMap implements CommandMap {
      */
     private synchronized boolean register(@NotNull String label, @NotNull Command command, boolean isAlias, @NotNull String fallbackPrefix) {
         knownCommands.put(fallbackPrefix + ":" + label, command);
-        if ((command instanceof BukkitCommand || isAlias) && knownCommands.containsKey(label)) {
+        // Paper start
+        Command known = knownCommands.get(label);
+        if ((command instanceof BukkitCommand || isAlias) && (known != null && !known.canBeOverriden())) {
+        // Paper end
             // Request is for an alias/fallback command and it conflicts with
             // a existing command or previous alias ignore it
             // Note: This will mean it gets removed from the commands list of active aliases
@@ -114,7 +125,9 @@ public class SimpleCommandMap implements CommandMap {
         // If the command exists but is an alias we overwrite it, otherwise we return
         Command conflict = knownCommands.get(label);
         if (conflict != null && conflict.getLabel().equals(label)) {
+            if (!conflict.canBeOverriden()) { // Paper
             return false;
+            } // Paper
         }
 
         if (!isAlias) {
